Support for readonly filesystems

mehtamohit013 opened the pull request:
## Describe your changes
Added support for read-only file systems

## Issue number

Closes #63 

## Checklist before requesting a review

- [X] Performed a self-review of my code
- [X] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [X] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [X] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)



<!-- readthedocs-preview ploomber-core start -->
----
:books: Documentation preview :books:: https://ploomber-core--67.org.readthedocs.build/en/67/

<!-- readthedocs-preview ploomber-core end -->

yafimvo reviewed (changes requested):
I think it looks good. added some comments.

@edublancas Do you think we should test it on an AWS instance to make sure it works?

tonykploomber reviewed (commented):
Overall LG, but can we have a real readonly folder to test?

mehtamohit013 commented on tests/telemetry/test_telemetry.py:
```diff
@@ -874,8 +876,7 @@ def test_permissions_error(monkeypatch):
     if is_read_only:
-        with pytest.raises(PermissionError):
-            telemetry.Internal()
+        telemetry.Internal()
```
@yafimvo 
- I modify the test `telemetry/test_telemetry.py::test_permissions_error`
- Added a new test: `test_config.py::test_readonly`

mehtamohit013 replied:
Not working on windows

mehtamohit013 commented:
> Overall LG, but can we have a real readonly folder to test?

Yup, I added it in `conftest.py` and testing: `test_config.py::test_readonly`

Passing for ubuntu and Macos. Working on fixing Windows one based on [this](https://stackoverflow.com/questions/12168110/how-to-set-folder-permissions-in-windows)

jinniw43805 commented:
Looks like a special fix for windows environment, try to see if it's easy fix, since this was happened originally in EC2 environment.

mehtamohit013 commented:
@edublancas @jinniw43805 @tonykploomber @yafimvo 

So, I tried to get it working on Windows based on [this](https://stackoverflow.com/questions/12168110/how-to-set-folder-permissions-in-windows) for [conftest.py](https://github.com/ploomber/core/pull/67/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128) , but it failed (Currently, I  put an `xfail` around it)
Needs a Windows VM for further debugging. 

Except the Windows part, resolved all the comments and PR is ready to merge

yafimvo commented on src/ploomber_core/telemetry/telemetry.py:
```diff
@@ -278,8 +280,9 @@ def check_dir_exist(input_location=None):
 
-    if not p.exists():
-        p.mkdir(parents=True)
+    if not (cls is not None and cls.writable is False):
```
Suggestion:
Instead of passing `cls` and checking if it's writable, we can add the parameter `create` with a default value of `True`.

Something like:

```python
def check_dir_exist(input_location=None, create=True):
    ...
    if not p.exists():
        if create:
            p.mkdir(parents=True)
    ...
```

now, we can do something like this

```python
@classmethod
def path(cls):
    conf_dir = check_dir_exist(CONF_DIR, create=cls.writable)
    return Path(conf_dir, DEFAULT_USER_CONF)
```

yafimvo reviewed (changes requested):
lgtm. added minor comments.

yafimvo commented:
> So, I tried to get it working on Windows based on [this](https://stackoverflow.com/questions/12168110/how-to-set-folder-permissions-in-windows) for [conftest.py](https://github.com/ploomber/core/pull/67/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128) , but it failed (Currently, I put an `xfail` around it) Needs a Windows VM for further debugging.
> 
> Except the Windows part, resolved all the comments and PR is ready to merge

Can we do something like this [test_permissions_error](https://github.com/ploomber/core/blob/main/tests/telemetry/test_telemetry.py#L862)?

I think if it doesn't work we can leave it for now since `test_permissions_error` is already checking this behavior.

mehtamohit013 commented:
@yafimvo 
> Can we do something like this [test_permissions_error](https://github.com/ploomber/core/blob/main/tests/telemetry/test_telemetry.py#L862)?

`os.chmod` doesn't works on windows. 
>Although Windows supports [chmod()](https://docs.python.org/2/library/os.html#os.chmod), you can only set the file’s read-only flag with it (via the stat.S_IWRITE and stat.S_IREAD constants or a corresponding integer value). All other bits are ignored.

I tried out setting using `stat.S_IWRITE` and `stat.S_IREAD`. It is not working

I tried out [this](https://stackoverflow.com/questions/12168110/how-to-set-folder-permissions-in-windows) thread, but it not working

yafimvo reviewed (changes requested):
Added one note.

Since we know internal and user configurations are working on windows (from [here](https://github.com/ploomber/core/blob/6011d153c870537b93367471335cbd97b61bd07e/tests/telemetry/test_telemetry.py#L863)), I think let's keep `def test_readonly(tmp_readonly_directory)` as is (skip on Windows) and remove the windows part (` if platform.system() == "Windows":`) from `def tmp_readonly_directory()`. Please add a short explanation of what's the issue on windows and add the links so we can fix it later.

mehtamohit013 commented on src/ploomber_core/telemetry/telemetry.py:
```diff
@@ -264,7 +266,7 @@ def get_home_dir():
 
 
-def check_dir_exist(input_location=None):
+def check_dir_exist(input_location=None, create=None):
```
Fixed

mehtamohit013 commented:
> Since we know internal and user configurations are working on windows (from [here](https://github.com/ploomber/core/blob/6011d153c870537b93367471335cbd97b61bd07e/tests/telemetry/test_telemetry.py#L863)), I think let's keep def test_readonly(tmp_readonly_directory) as is (skip on Windows) and remove the windows part ( if platform.system() == "Windows":) from def tmp_readonly_directory(). Please add a short explanation of what's the issue on windows and add the links so we can fix it later.

I believe the function is correct, just needs some little debugging. So, I am keeping the function there for future reference. 
Added a `#TODO` comment.
Also, I added the links in the if-else statement.

yafimvo commented:
> I believe the function is correct, just needs some little debugging. So, I am keeping the function there for future reference.
> Added a `#TODO` comment.
> Also, I added the links in the if-else statement

@mehtamohit013 Can you please open an issue so it won't get lost?

edublancas commented on setup.py:
```diff
@@ -19,13 +21,10 @@
+DEV = ["pytest", "flake8", "twine", "invoke", "pkgmt"]
+
+if platform.system() == "Windows":
+    DEV.append("pywin32")
```
this won't work for built distributions (.whl) ,we can  add add a  new section in extras_require to account for this: https://stackoverflow.com/a/6470373/709975

edublancas replied:
yeah, that sounds good

edublancas commented on src/ploomber_core/config.py:
```diff
@@ -112,19 +116,46 @@ def __setattr__(self, name, value):
 
         return config
 
+    def random_string(self, length=10):
```
no need to create a separate method if we're only using this to test if the filesystem is writable, we can embed the logic in filesystem_writable

edublancas commented on src/ploomber_core/config.py:
```diff
@@ -112,19 +116,46 @@ def __setattr__(self, name, value):
+        sequence = f"{string.ascii_uppercase}{string.digits}"
+        return "".join(random.choices(sequence, k=length))
+
+    def filesystem_writable(self):
```
```suggestion
    def _filesystem_writable(self):
```

let's make this a private method

edublancas commented on src/ploomber_core/config.py:
```diff
@@ -112,19 +116,46 @@ def __setattr__(self, name, value):
+
+    def filesystem_writable(self):
+        """Check if the filesystem is writable"""
+        try:
+            # random string is used for race conditions when using multiprocessing
+            tmp = self.path().parent / f"tmp_{self.random_string()}.txt"
+            tmp.touch()
+            tmp.unlink()
+            return True
+        except PermissionError:
+            return False
```
it's a best practice to have a single line in the try statement:



```python
        try:
            tmp.touch()
            return True
        except PermissionError:
            return False
```

you can move the other lines outside the try statement.

and you can move the unlink in the `else` statement (after the except statement)

edublancas commented on src/ploomber_core/config.py:
```diff
@@ -112,19 +116,46 @@ def __setattr__(self, name, value):
+            tmp.unlink()
+            return True
+        except PermissionError:
+            return False
```
yeah, I think we should be logging all exceptions (we can use the logging module)

edublancas commented on src/ploomber_core/telemetry/telemetry.py:
```diff
@@ -278,8 +280,9 @@ def check_dir_exist(input_location=None):
 
-    if not p.exists():
-        p.mkdir(parents=True)
+    if create:
+        if not p.exists():
```
this can be simplified in a single if statement

edublancas commented on src/ploomber_core/telemetry/telemetry.py:
```diff
@@ -66,10 +66,11 @@ class UserSettings(Config):
     cloud_key: str = None
     user_email: str = None
     stats_enabled: bool = True
+    writable: bool = True
```
do we need to add this property here? this will add a new property to the stored file, which users can modify. but we only need this for internal purposes, so it looks like we're exposing internal stuff. if this is a subclass of Config, can't we use the existing property declared in the subclass?

edublancas replied:
So I see you added writable to Config, that's fine.

However, you've also added it to all the subclasses (e.g., Internal). My point is that we should not add it as an annotation to the subclasses because all annotations are persistent to the YAML file (you can see the Config implementation and you'll see that __annotations__ is stored in the YAML file) because it's exposing internals of the implementation.

So the end goal is: Config should manage its internal state transparently. subclasses of config should no need any modifications because that's the point of making them subclasses.

What happens if we delete the annotation? if Config defines writable, subclasses should have access to it, right?

Also, let's rename it to `_writable` to indicate that this is an internal method, not part of the public API.

edublancas commented:
> What kind of logging do you want to add?
Currently added :

yes. this logging.warning makes sense

edublancas reviewed (changes requested):
left some comments

mehtamohit013 commented:
@edublancas 
A few points:
- Removed dependency on [`check_dir_exist`](https://github.com/ploomber/core/blob/e276eea63b502208c312b1c6f40db9513f4a43f7/src/ploomber_core/telemetry/telemetry.py#L267) function, as the code is stuck in an infinite loop in some cases. Added functionality to `config.py` to automatically create dir if not exists.
- Minor: While debugging, I thought there was some error when setting attributing `DEFAULT_PLOOMBER_DIR` to `.`. So, replaced it with pathlib function.

edublancas reviewed (changes requested):
I tested it and it works well. However, I noticed that Python logs `.warning()` by default, hence users will see this:

```
(environment_name) user@b231cfac945f:~$ ploomber-engine sum.ipynb out.ipynb
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
Executing cell: 2: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 786.78it/s]
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
WARNING:root:Filesystem is not writable. Telemetry won't be saved
```

from a user's perspective, they don't care if telemetry is working or not. so let's remove the logging calls. I think we can add them later if we need to debug them.

in case you need to test things, this is how I did it:

```sh
docker run -it condaforge/mambaforge /bin/bash

# create a new user and make the ~/.ploomber dir read-only
adduser user
mkdir -p /home/user/.ploomber
chmod -R 555 /home/user/.ploomber

# switch to the user
su - user

# ensure this is read only
touch /home/user/.ploomber/a

# create a new env
conda create --name env python=3.10 -y
conda activate env


# get sample notebook and ploomber-engine
wget https://raw.githubusercontent.com/ploomber/ploomber-engine/main/examples/sum.ipynb
pip install ploomber-engine

# this should fail
ploomber-engine sum.ipynb out.ipynb

# install version with read-only fix
pip uninstall ploomber-core -y
pip install git+https://github.com/mehtamohit013/core@readonly

# this should work
ploomber-engine sum.ipynb out.ipynb
```

mehtamohit013 commented:
@edublancas 
Removed logging and tested it using the script above

Ready for merge.
